sequencer/src/block/full_payload/ns_table.rs

```diff
@@ -166,17 +204,58 @@ impl NsTable {
 
     /// Does the `index`th entry exist in the namespace table?
     pub fn in_bounds(&self, index: &NsIndex) -> bool {
-        // The number of entries in the namespace table, including all duplicate
-        // namespace IDs.
-        let num_nss_with_duplicates = std::cmp::min(
-            // Number of namespaces declared in the ns table
-            self.read_num_nss(),
-            // Max number of entries that could fit in the namespace table
-            self.bytes.len().saturating_sub(NUM_NSS_BYTE_LEN)
-                / NS_ID_BYTE_LEN.saturating_add(NS_OFFSET_BYTE_LEN),
-        );
+        self.len().in_bounds(index)
+    }
+
+    /// Are the bytes of this [`NsTable`] uncorrupted?
+    ///
+    /// # Checks
+    /// 1. Byte length must hold a whole number of entries.
+    /// 2. All namespace IDs and offsets must increase monotonically. Offsets
+    ///    must be nonzero.
+    /// 3. Header consistent with byte length (obsolete after
+    ///    <https://github.com/EspressoSystems/espresso-sequencer/issues/1604>)
+    pub fn validate(&self) -> Result<(), NsTableValidationError> {
+        use NsTableValidationError::*;
+
+        // Byte length for a table with `x` entries must be exactly
+        // `x * NsTableBuilder::entry_byte_len() + NsTableBuilder::header_byte_len()`
+        if self.bytes.len() % NsTableBuilder::entry_byte_len() != NsTableBuilder::header_byte_len()
```

Comment:
Nit: this would fail if `header_byte_len` is a multiple of `entry_byte_len`

Reply:
Fixed in https://github.com/EspressoSystems/espresso-sequencer/pull/1607/commits/1807a243f96a5ddcf45f8276da07beb1e7a2eb83.

That is not a nit; it's a latent bug. Maybe one day I'll post a PR where you don't find a bug in review. 🤦‍♂️

Reply:
Fixed in https://github.com/EspressoSystems/espresso-sequencer/pull/1607/commits/e1ef35f86a979648bae18329e22fa24533b2f567.

That's not a nit; it's a latent bug. Perhaps one day I'll post a PR where you don't find a bug in review. 🤦‍♂️